Automattic/vip-go-mu-plugins
# Limit calls to `Two_Factor_Core::is_user_using_two_factor()`

## Description

There has been a report of an issue with users being unable to edit any user profile field except 2FA fields on /wp-admin/profile.php. While the 2FA plugin is enabled, users were seeing an `Invalid User ID` error while trying to update their `First Name` profile field. We reproduced this on other fields as well although updating one of the 2FA fields always worked. Disabling the 2FA plugin with add_filter( 'wpcom_vip_enable_two_factor', '__return_false' ); did solve the issue and helped us isolate the problem to something dealing with the 2FA plugin. The full steps to reproduce the issue can be found in #1444.

First, a bit of the debugging that took place on this issue:
I traced a profile edit from form submission all the way up to `wp_insert_user()`. If `clean_user_cache()` [here](https://github.com/WordPress/WordPress/blob/c3f9af4f00375d03bf4a4be5185e5c09abca094d/wp-includes/user.php#L1986) is not called, the data gets through just fine. If it is called, the user cache is deleted and when `wp_insert_user()` is called, it’s going to call `get_userdata( $ID )`. That eventually calls `WP_User::get_data_by( $field, $value )` passing in `ID` as the field and the user ID as the value.

`WP_User::get_data_by()` looks in the cache for the user data (which doesn’t exist because `clean_user_cache()` deleted it) with `wp_cache_get( $user_id, 'users' )`. Since nothing is in the cache for that user, it tries to get it directly from the database with `$wpdb->get_row( $wpdb->prepare( "SELECT * FROM $wpdb->users WHERE ID = %s LIMIT 1", $value ) );`. Running that query just once results in `null` being returned. Running it again immediately, however, returns the proper result. This lead us to believe that this could be a db lag issue.

I also noticed that if I disabled the `map_meta_cap` filter [here](https://github.com/Automattic/vip-go-mu-plugins/blob/98ca551386d18f5f200dfe146d7626742b6c2bf8/two-factor.php#L177), profile fields could be updated without issue. I briefly went down the path of this potentially being an issue with how we were restricting capabilities but realized that wasn't the core issue. But I did realize that the function hooked to that filter called `wpcom_vip_is_two_factor_forced()` each time for every capability.

Next, I started looking at each function called by `wpcom_vip_is_two_factor_forced()` and saw that `Two_Factor_Core::is_user_using_two_factor()` results in several user-related WordPress core functions being called. These could be causing a lot of database writes and may have lead to a database replication lag issue. If there was a replication lag, direct queries for a user could be returning zero results and cause profile updates fail with an `Invalid User ID` error.

This change stores the result of `Two_Factor_Core::is_user_using_two_factor()` (by user ID) in a static variable and only calls it once per user. Doing this, user profile updates are always successful because a user is always found in the database.

This issue is **not** reproducible on a VIP Go sandbox (without a few mu-plugins edits) OR while connected to the Automattic proxy.

## Checklist

Please make sure the items below have been covered before requesting a review:


## Steps to Test

To test this on a VIP sandbox, you must make edits to a few files in `mu-plugins/1/`: 

```
diff --git a/000-vip-init.php b/000-vip-init.php
index 1f9e4211..5463fed2 100644
--- a/000-vip-init.php
+++ b/000-vip-init.php
@@ -72,7 +72,7 @@ if ( ! defined( 'WPCOM_IS_VIP_ENV' ) ) {
        define( 'WPCOM_IS_VIP_ENV', false );
 }

-define( 'WPCOM_SANDBOXED', false !== strpos( gethostname(), '_web_dev_' ) );
+define( 'WPCOM_SANDBOXED', false );
 define( 'VIP_GO_IS_CLI_CONTAINER', false !== strpos( gethostname(), '_wpcli_' ) );

 // Used to verify emails sent via our SMTP servers
@@ -143,7 +143,7 @@ define( 'CRON_CONTROL_ADDITIONAL_INTERNAL_EVENTS', $internal_cron_events );
 // Avoids issues with `get_filesystem_method` which attempts to write to `WP_CONTENT_DIR` and fails.
 define( 'FS_METHOD', 'direct' );

-if ( WPCOM_SANDBOXED ) {
+if ( false !== strpos( gethostname(), '_web_dev_' ) ) {
        require __DIR__ . '/vip-helpers/sandbox.php';
 }
```

```
diff --git a/vip-helpers/sandbox.php b/vip-helpers/sandbox.php
index 49935635..be9d8c4c 100644
--- a/vip-helpers/sandbox.php
+++ b/vip-helpers/sandbox.php
@@ -1,8 +1,8 @@
 <?php

-if ( ! WPCOM_SANDBOXED ) {
-	return;
-}
+// if ( ! WPCOM_SANDBOXED ) {
+// 	return;
+// }

 add_action( 'wp_footer', 'wpcom_do_sandbox_bar', 100 );
 add_action( 'admin_footer', 'wpcom_do_sandbox_bar', 100 );
```

1. After you've made the above edit to your sandbox, and verified that you are not connected to proxy, you should be able to reproduce this issue on your VIP Go sandbox.
1. Verify that updating a profile field results in an `Invalid User ID` error. You may find the first attempt is successful, but subsequent submissions should fail.
1. Apply the changes in this PR.
1. Update a user profile field and verify that the change was successful.

Fixes #1444
